pinctrl: add nrf implementation

gmarull opened the pull request:
This PR adds an initial implementation of a pinctrl driver for nRF based on the pinctrl API proposal in #37572. The driver only supports the UART/E peripherals for now, but could be extended to cover the rest. A sample on how dynamic pinctrl can be used is also provided.

Note: it will be enabled and boards ported once all peripherals are ready.

gmarull committed:
drivers: pinctrl: nrf: initial support

Add initial support for nRF pin controller driver. The implementation in
this patch does not yet support any peripheral. Only states
representation and basic driver functionality is introduced.

Note:
The nrf_pin_configure function has been marked as __unused since it may
not be used in certain scenarios until all peripherals are supported by
the pinctrl driver. For example, if only UART/E is supported but the
board does not enable UART, the function will never get called. However,
that board will likely have other peripherals that will gain support in
the future.

Thanks to Marti Bolivar for bindings documentation.

Co-authored-by: Marti Bolivar <marti.bolivar@nordicsemi.no>
Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pinctrl: nrf: add support for uart/uarte peripheral

Add support for configuring UART/UARTE peripheral pins.

Co-authored-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>
Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: nrfx_uarte: add support for pinctrl

This patch adds support for the new pinctrl API to the UARTE driver. The
old pin property based solution is still kept so that users have time to
transition to the new model.

Notes:

- Some build assertions cannot be performed since the driver does not
  have direct access to pin settings anymore. As a result user will not
  be notified if HWFC is enabled but RTS/CTS pins are not configured.
- Hardware flow control can be enabled regardless of pin configuration,
  it is now up to the user to configure RTS/CTS pins in DT.
- Some RX enable checks that were performed using pin information has
  been replaced with a DT property that informs if RX is enabled or not.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: drivers: pinctrl: add test for checking nRF DT parsing

Add a unit test that allows to validate if pin control configuration
stored in Devicetree is parsed correctly.

nRF boards used to run this test should add the pinctrl:nrf tag.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: gpio: nrfx: adjust initialization level and priority

Make nrfx GPIO driver part of the PRE_KERNEL_1 initialization stage. As
a result, the GPIO driver can now be initialized before UART if
required, a device that is also initialized during PRE_KERNEL_1.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: boards: nrf: dynamic_pinctrl: initial version

Add a sample that shows how dynamic pin control can be useful to remap a
peripheral to a different set of pins.

NOTE: default pin configuration is provided for the nRF52840-DK board.
Such configuration will be removed once all nRF-based boards are ported
to pinctrl.

Thanks to Francesco for documentation suggestions.

Co-authored-by: Francesco D. Servidio <francesco.servidio@nordicsemi.no>
Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: nrfx_uart: add support for pinctrl

This patch adds support for the new pinctrl API to the UART driver. The
old pin property based solution is still kept so that users have time to
transition to the new model.

Notes:

- A new property to disable RX has been introduced: disable-rx. It is no
  longer possible to do it automatically depending on pin information,
  since it's not available when using pinctrl.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
scripts: utils: add pinctrl migration script for nRF boards

This script can be used to automatically migrate the Devicetree files of
nRF-based boards using the old <signal>-pin properties to select
peripheral pins. The script will parse a board Devicetree file and will
first adjust that file by removing old pin-related properties replacing
them with pinctrl states.  A board-pinctrl.dtsi file will be generated
containing the configuration for all pinctrl states. Note that script
will also work on files that have been partially ported.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented on drivers/pinctrl/pinctrl_nrf.c:
```diff
@@ -28,10 +36,69 @@ static void nrf_pin_configure(pinctrl_soc_pin_t pin, nrf_gpio_pin_dir_t dir,
 	switch (reg) {
+#if defined(NRF_UART_TYPE)
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(uart0), okay)
+	case DT_REG_ADDR(DT_NODELABEL(uart0)):
```
@anangl implemented your suggestion (and added you to the sign-off, since it's really your code :-) ), that's indeed a good solution IMO.

gmarull commented on boards/arm/96b_carbon_nrf51/96b_carbon_nrf51-pinctrl.dtsi:
```diff
@@ -0,0 +1,22 @@
+&pinctrl {
+	uart0_default: uart0_default {
+		pins1 {
+			pins = <NRF_PIN(0, 29, UART_TX)>,
+			       <NRF_PIN(0, 11, UART_RX)>;
```
another naming choice could be:
```
psels = <NRF_PSEL(0, 29, UART_TX)>,
        <NRF_PSEL(0, 11, UART_RX)>;
```

@carlescufi @anangl @mbolivar-nordic

mbolivar-nordic replied:
> `psels = <NRF_PSEL(0, 29, UART_TX)>,`

Interesting. If we are going to do it like this, maybe it would be worthwhile to change the order of the arguments so it looks like `NRF_PSEL(UART_TX, 0, 29)` (etc.) instead.

Just an idea because the product specification's document PSEL.TXD etc., so putting the pin function (TX) next to the `PSEL` token in DTS matches that order.

anangl replied:
Looks reasonable. I think the internal macros in `pinctrl_nrf.c` could also use different naming: `NRF_PSEL_UART` instead of `NRF_UART_PSEL`.

gmarull replied:
I've renamed macros to follow `NRF_PSEL_XXX` naming

mbolivar-nordic replied:
So does that mean you're dropping the `psels` property idea?

gmarull replied:
not yet, I'd like to hear opinions @carlescufi @anangl ?

anangl replied:
I thought I had already expressed mine above. I like this idea, @mbolivar-nordic's suggestion to change the order, too.

carlescufi replied:
Yes, I do like that idea. It is consistent with the datasheets and it describes better the purpose of the macro, which is to associate a function's signal to a pin number. Let us go with that then, really nice idea.

gmarull replied:
updated PR

mbolivar-nordic commented on dts/bindings/pinctrl/nordic,nrf-pinctrl.yaml:
```diff
@@ -0,0 +1,70 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    nRF pin controller
```
I think the major things to cover here are:

- this is a singleton that is already provided for every SoC
- it has a node label named `pinctrl` you can use to set properties
- it should be changed in board files and applications to contain a global set of pin configurations for all peripherals
- example pin configurations (e.g. UART)

I know this is just a draft, and I'm not asking for this to be done right now. Just making a note for later.

gmarull replied:
I've re-written the description, I think it now contains all the points you mention. Let me know if it's better now.

mbolivar-nordic commented on dts/bindings/pinctrl/nordic,nrf-pinctrl.yaml:
```diff
@@ -0,0 +1,70 @@
+
+  child-binding:
+    description: |
+      nRF pin controller pin configuration group. Each group contains a list of
```
I think this may be clearer if it appears in the parent description (the one for the child binding, not the grandchild binding).

gmarull replied:
moved all details to the parent description

mbolivar-nordic commented on soc/arm/nordic_nrf/common/CMakeLists.txt:
```diff
@@ -2,3 +2,4 @@
 # SPDX-License-Identifier: Apache-2.0
 
 zephyr_library_sources_ifdef(CONFIG_SOC_FAMILY_NRF soc_nrf_common.S)
+zephyr_include_directories(.)
```
Might be a bit cleaner to move `../soc_nrf_common.h` to this directory, and `include <soc_nrf_common.h>` from each of the `../nrf*/soc.h` files, if you're adding a new common include directory here. Just an idea; feel free to ignore.

gmarull replied:
done in a new commit

mbolivar-nordic commented on scripts/utils/pinctrl_nrf_migrate.py:
```diff
@@ -0,0 +1,539 @@
+        };
+    };
+
+Copyright (c) 2021 Nordic Semiconductor ASA
+SPDX-License-Identifier: Apache-2.0
```
Move to comment instead of docstring?

gmarull replied:
done

mbolivar-nordic commented on tests/drivers/pinctrl/nrf/boards/nrf52840dk_nrf52840.overlay:
```diff
@@ -0,0 +1,42 @@
+/*
```
Why bother limiting this to nRF52840-DK?

It's already in an nrf directory, and this should work on all SoCs, so you could just make this app.overlay and allow testing it on any board a person has available -- importantly, including nrf52_bsim.

gmarull replied:
moved to app.overlay. I'm still filtering with `platform_allow`, though. I'm not sure what would be the best way to allow this test to run on all nRF based targets. Filtering by Kconfig is not an option, since config step would fail due to the app.overlay content.

mbolivar-nordic replied:
Can't you use integration_platforms to specify this should be built with board 'foo', without limiting to just 'foo'? IIRC that's the difference between platform_allow and integration_platforms.

gmarull replied:
good point, I have updated to `integration_platforms`.

gmarull replied:
@mbolivar-nordic CI tries to build for `native_posix` now (https://github.com/zephyrproject-rtos/zephyr/runs/4190937903?check_suite_focus=true), suggestions?

mbolivar-nordic replied:
I'm out of my element, that's not how I thought twister was supposed to work. @PerMac can you help us restrict the twister run to just the platforms in integration_platforms for this PR?

@gmarull maybe `filter: dt_compat_enabled("nordic,nrf-pinctrl")` will help?

gmarull replied:
It will not work, filter relies on the DT pickle file generated at configuration time, a step that will fail (due to DT syntax errors, `NRF_PIN` not available).

gmarull replied:
A solution may be to rely on tags

mbolivar-nordic replied:
CONFIG_SOC_COMPATIBLE_NRF?

gmarull replied:
same problem: it relies on CMake config step to succeed.

PerMac replied:
I suggest adding `platform_exclude: native_posix` to satisfy the CI. The gh action that failed (clang-build) has native_posix explicitly given in the twister call there. It does not use `integration_platforms` nor `--integration` (integration mode). `Integration_platforms` are used only when `--integration` is used. 
I know that platform selection in CI is hard to follow but until we upgrade twister there probably is no better way for this. 
As @gmarull pointed out: we can only use 'yaml parsing' ways to filter in this case. Using dt or kconfig filtering is only possible after initial cmake passes.

mbolivar-nordic reviewed (commented):
Looking really awesome. A few random comments. This is great work.

gmarull committed:
soc: arm: nordic_nrf: move soc_nrf_common.h to common folder

Since the addition of pinctrl the common folder is added to the Zephyr
include path. This can be re-used to place the soc_nrf_common.h header
in the common folder and include it directly, without the need of
relative paths.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented on soc/arm/nordic_nrf/common/pinctrl_soc.h:
```diff
@@ -0,0 +1,114 @@
+ * @param prop Property name describing state pins.
+ */
+#define Z_PINCTRL_STATE_PINS_INIT(node_id, prop)			       \
+	{DT_FOREACH_CHILD_VARGS(DT_PROP_BY_IDX(node_id, prop, 0),	       \
```
moved to `DT_PHANDLE`, makes sense and makes code more readable.

mbolivar-nordic commented on dts/bindings/pinctrl/nordic,nrf-pinctrl.yaml:
```diff
@@ -0,0 +1,93 @@
+
+    List of supported standard properties:
+
+    - bias-disable: Disable pull-up/down (default behavior, not required).
+    - bias-pull-up: Enable pull-up resistor.
+    - bias-pull-down: Enable pull-down resistor.
+    - low-power-enable: Configure pin as an input with input buffer
+      disconnected.
+
+    Note that bias options are mutually exclusive.
+
+    Usage example:
+
+    /* board.dts */
+    #include "board-pinctrl.dtsi"
+
+    uart0 {
+      pinctrl-0 = <&uart0_default>;
+      pinctrl-names = "default";
+    };
+
+    /* board-pinctrl.dtsi */
+
+    &pinctrl {
+      /* configuration for uart0 device, default state */
+      uart0_default: uart0_default {
+        /* group 1 ('pins1' name is arbitrary) */
+        pins1 {
+          /* configure P0.1 as UART_TX and P0.2 as UART_RTS */
+          pins = <NRF_PIN(0, 1, UART_TX)>, <NRF_PIN(0, 2, UART_RTS)>;
+        };
+        /* group 2 */
+        pins2 {
+          /* configure P0.3 as UART_RX and P0.4 as UART_CTS */
+          pins = <NRF_PIN(0, 3, UART_RX)>, <NRF_PIN(0, 4, UART_CTS)>;
+          /* both P0.3 and P0.4 are configured with pull-up */
+          bias-pull-up;
+        };
+      };
+    };
```
```suggestion
    The nRF pin controller is a singleton node responsible for controlling
    pin function selection and pin properties. For example, you can use this
    node to route UART0 RX to pin P0.1 and enable the pull-up resistor on the
    pin.

    The node has the 'pinctrl' node label set in your SoC's devicetree,
    so you can modify it like this:

      &pinctrl {
              /* your modifications go here */
      };

    All device pin configurations should be placed in child nodes of the
    'pinctrl' node, as shown in this example:

      /* You can this in places like a board-pinctrl.dtsi file in
       * your board directory, or a devicetree overlay in your application.
       */
      &pinctrl {
        /* configuration for uart0 device, default state */
        uart0_default: uart0_default {
          /* group 1 ('pins1' name is arbitrary) */
          pins1 {
            /* configure P0.1 as UART_TX and P0.2 as UART_RTS */
            pins = <NRF_PIN(0, 1, UART_TX)>, <NRF_PIN(0, 2, UART_RTS)>;
          };
          /* group 2 */
          pins2 {
            /* configure P0.3 as UART_RX and P0.4 as UART_CTS */
            pins = <NRF_PIN(0, 3, UART_RX)>, <NRF_PIN(0, 4, UART_CTS)>;
            /* both P0.3 and P0.4 are configured with pull-up */
            bias-pull-up;
          };
        };
      };

    The 'uart0_default' child node encodes the pin configurations for a
    particular state of a device; in this case, the default (that is, active)
    state. You would specify the low-power configuration for the same device
    in a separate child node.

    As shown, pin configurations are organized in groups within each child node.
    Each group can specify a list of pin function selections in the 'pins' property.

    A group can also specify shared pin properties common to all the specified
    pins, such as the 'bias-pull-up' property in group 2. Here is a list of supported
    standard pin properties:

    - bias-disable: Disable pull-up/down (default behavior, not required).
    - bias-pull-up: Enable pull-up resistor.
    - bias-pull-down: Enable pull-down resistor.
    - low-power-enable: Configure pin as an input with input buffer
      disconnected.

    Note that bias options are mutually exclusive.

    To link this pin configuration with a device, use a pinctrl-N property
    for some number N, like this example you could place in your board's DTS
    file:

       #include "board-pinctrl.dtsi"

       uart0 {
             pinctrl-0 = <&uart0_default>;
             pinctrl-names = "default";
       };
```

Looks good, just some suggested edits

mbolivar-nordic replied:
cc @FrancescoSer

gmarull commented:
> I'm out of my element, that's not how I thought twister was supposed to work. @PerMac can you help us restrict the twister run to just the platforms in integration_platforms for this PR?

> @gmarull maybe `filter: dt_compat_enabled("nordic,nrf-pinctrl")` will help?

It will not work, filter relies on the DT pickle file generated at configuration time, a step that will fail (due to DT syntax errors, `NRF_PIN` not available).

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -1145,18 +1173,44 @@ static void uart_nrfx_pins_enable(const struct device *dev, bool enable)
 static int uart_nrfx_pm_action(const struct device *dev,
 			       enum pm_device_action action)
 {
+	const struct uart_nrfx_config *config = get_dev_config(dev);
```
When `CONFIG_PINCTRL` is not defined, an "unused variable" warning may be generated for `config`. I think you could pass this variable as a parameter to `uart_nrfx_pins_configure` to prevent the warning (this would save `#ifdef CONFIG_PINCTRL` here and in `uart_nrfx_init()`).

gmarull replied:
I prefer to use `dev` as an argument, so guarded config using ifdef.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -189,6 +188,48 @@ static inline NRF_UARTE_Type *get_uarte_instance(const struct device *dev)
+	const struct uarte_nrfx_config *cfg = get_dev_config(dev);
+
+	if (!sleep) {
+		nrf_gpio_pin_write(cfg->tx_pin, 1);
+		nrf_gpio_cfg_output(cfg->tx_pin);
```
This assumes that `tx_pin` has a valid value (the `tx-pin` property was required so far), but currently it can be `NRF_UARTE_PSEL_DISCONNECTED`. Similar story in `uart_nrfx_uart.c`.

gmarull replied:
guarded against disconnected possibility in both uart/e drivers.

gmarull commented on drivers/pinctrl/pinctrl_nrf.c:
```diff
@@ -0,0 +1,91 @@
+	      (NRF_DRIVE_H0H1 == NRF_GPIO_PIN_H0H1) &&
+	      (NRF_DRIVE_D0S1 == NRF_GPIO_PIN_D0S1) &&
+	      (NRF_DRIVE_D0H1 == NRF_GPIO_PIN_D0H1) &&
+	      (NRF_DRIVE_D0S1 == NRF_GPIO_PIN_D0S1) &&
```
fixed

gmarull commented on drivers/pinctrl/pinctrl_nrf.c:
```diff
@@ -0,0 +1,63 @@
+	      (NRF_DRIVE_H0D1 == NRF_GPIO_PIN_H0D1) &&
+#if defined(GPIO_PIN_CNF_DRIVE_E0E1)
+	      (NRF_DRIVE_E0E1 == NRF_GPIO_PIN_E0E1) &&
+#endif /* defined(GPIO_PIN_CNF_DRIVE_E0D1) */
```
fixed, left the comment though (I prefer to add it to all if/defs for consistency)

gmarull commented on dts/bindings/serial/nordic,nrf-uart-common.yaml:
```diff
@@ -68,12 +36,56 @@ properties:
+            tx-pin = <16>;
+        For pins P1.0 through P1.31, add 32 to the pin number. For
+        example, to use P1.2 for TX, set:
+            tx-pin = <34>;  /* 32 + 2 */
```
added them back, reordering mistake

gmarull commented on samples/boards/nrf/dynamic_pinctrl/README.rst:
```diff
@@ -0,0 +1,85 @@
+Building and Running
+********************
+
+This application can be built for nRF52840DK as follows:
```
fixed

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -1145,18 +1177,45 @@ static void uart_nrfx_pins_enable(const struct device *dev, bool enable)
 static int uart_nrfx_pm_action(const struct device *dev,
 			       enum pm_device_action action)
 {
+#ifdef CONFIG_PINCTRL
+	const struct uart_nrfx_config *config = get_dev_config(dev);
+#endif
+	int ret;
```
```suggestion
#ifdef CONFIG_PINCTRL
	const struct uart_nrfx_config *config = get_dev_config(dev);
	int ret;
#endif
```

gmarull replied:
done

anangl commented on dts/bindings/serial/nordic,nrf-uart-common.yaml:
```diff
@@ -7,73 +7,88 @@ properties:
+        - 1200
+        - 2400
+        - 4800
+        - 9600
+        - 14400
+        - 19200
+        - 28800
+        - 31250
+        - 38400
+        - 56000
+        - 57600
+        - 76800
+        - 115200
+        - 230400
+        - 250000
+        - 460800
+        - 921600
+        - 1000000
+
     tx-pin:
       type: int
-      required: true
       description: |
-        The TX pin to use.
+        DEPRECATED: This option will only be used if the new pin control driver
+        is not enabled.
 
+        The TX pin to use.
         For pins P0.0 through P0.31, use the pin number. For example,
         to use P0.16 for TX, set:
-
             tx-pin = <16>;
-
         For pins P1.0 through P1.31, add 32 to the pin number. For
         example, to use P1.2 for TX, set:
-
             tx-pin = <34>;  /* 32 + 2 */
```
In https://github.com/zephyrproject-rtos/zephyr/pull/39429#discussion_r749174625 I meant these three removed empty lines, too. Removing them affects appearance of the bindings documentation, so I think they should be also restored.

gmarull replied:
sorry, I misunderstood that part, fixed

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -189,6 +188,52 @@ static inline NRF_UARTE_Type *get_uarte_instance(const struct device *dev)
+		}
+
+		if (cfg->rx_pin != NRF_UARTE_PSEL_DISCONNECTED) {
+			nrf_gpio_cfg_input(cfg->rx_pin, cfg->rx_pull_up);
```
`nrf_gpio_cfg_input()` needs a `nrf_gpio_pin_pull_t` as the second parameter and `cfg->rx_pull_up` is a `bool`. Same story with `cts_pull_up` below (and in `uart_nrfx_uart.c`).

gmarull replied:
fixed

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -1145,18 +1177,45 @@ static void uart_nrfx_pins_enable(const struct device *dev, bool enable)
+
 	switch (action) {
 	case PM_DEVICE_ACTION_RESUME:
-		uart_nrfx_pins_enable(dev, true);
```
Body of `uart_nrfx_pins_enable()` should be removed, too.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -1838,10 +1829,23 @@ static int uarte_nrfx_pm_action(const struct device *dev,
 	struct uarte_nrfx_data *data = get_dev_data(dev);
 #endif
+	const struct uarte_nrfx_config *cfg = get_dev_config(dev);
+	int ret;
```
This variable is not used when `CONFIG_PINCTRL` is not defined.

FrancescoSer commented on samples/boards/nrf/dynamic_pinctrl/README.rst:
```diff
@@ -0,0 +1,85 @@
+Sample Output
+=============
+
+In order to test the two configurations, connect a USB to UART adapter to both
+set of pins. If the board routes the default configuration to a virtual COM port
+such as in nRF52840 DK you can directly use that port.
+
+First of all open two serial terminals, one connected to the default set of pins
+and the other connected to the alternative set of pins.
+
+.. figure:: images/terminals-empty.png
+
+    Two serial terminals (left: default, right: alternative)
+
+Then, turn the board on. You should see a ``Hello World!`` message printed on
+the first terminal.
+
+.. figure:: images/terminals-default.png
+
+    ``Hello World!`` printed on the default set of pins
+
+Finally, press and hold the configuration selection push button (Button 1 on
+nRF52840 DK) and press the board reset button. You now should see a
+``Hello World!`` message on the second terminal.
+
+.. figure:: images/terminals-alt.png
+
+    ``Hello World!`` printed on the alternative set of pins
```
```suggestion
To test the two configurations follow these steps:

1. Connect a USB-to-UART adapter to both sets of pins. If the board routes the
   default configuration to a virtual COM port (as in the nRF52840 DK), you can
   directly use that port.

#. Open two serial terminals, one connected to the default set of pins and the
   other connected to the alternative set of pins.

   .. figure:: images/terminals-empty.png

       Two serial terminals (left: default, right: alternative).

#. Turn the board on. You should see a ``Hello World!`` message printed on
   the first terminal.

   .. figure:: images/terminals-default.png

       ``Hello World!`` printed on the default set of pins.

#. Press and hold the configuration-selection push-button (button 1 on the
   nRF52840 DK) and press the board reset button. You now should see a
   ``Hello World!`` message on the second terminal.

   .. figure:: images/terminals-alt.png

       ``Hello World!`` printed on the alternative set of pins.
```

FrancescoSer commented on samples/boards/nrf/dynamic_pinctrl/README.rst:
```diff
@@ -0,0 +1,85 @@
+nRF52840 DK
+===========
+
+The diagram below shows the pins assigned to the default and alternative
+configurations. If the user powers the board on the ``uart0`` peripheral will
+be routed to the default set of pins. Note that the default set of pins is also
+exposed via the left USB connector as a virtual COM port. If the user powers
+the board on while holding Button 1, the ``uart0`` peripheral will be routed to
+the alternative set of pins.
+
+.. figure:: images/nrf52840dk-dynamic-pinctrl.png
+
+    Configuration for nRF52840 DK
```
```suggestion
The diagram below shows the pins assigned to the default and alternative
configurations.

.. figure:: images/nrf52840dk-dynamic-pinctrl.png

    Configuration for nRF52840 DK

If you power on the board, the ``uart0`` peripheral is routed to the default
set of pins. The default set of pins is also exposed through the left USB
connector as a virtual COM port.

If you power on the board while holding Button 1, the ``uart0`` peripheral is
routed to the alternative set of pins.
```

FrancescoSer reviewed (changes requested):
Just a small review of the docs

gmarull commented on dts/bindings/pinctrl/nordic,nrf-pinctrl.yaml:
```diff
@@ -0,0 +1,107 @@
+          be defined using the NRF_PSEL utility macro that encodes the port,
+          pin and function.
+
+      drive-mode:
```
@mbolivar-nordic I wonder if this non-standard property should be prefixed with vendor name, similar to e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/timer/st%2Cstm32-timers.yaml#L23 is there any established policy?

mbolivar-nordic replied:
Zephyr does not generally follow the Linux practice of prefixing any vendor-specific properties with the vendor prefix itself. My guess is the STM32 maintainers are following along with patterns they already have in their Linux bindings.

gmarull commented on boards/arm/96b_carbon_nrf51/96b_carbon_nrf51-pinctrl.dtsi:
```diff
@@ -0,0 +1,22 @@
+
+&pinctrl {
+	uart0_default: uart0_default {
+		pins1 {
```
should we use `group1..N` now? or is `pins1..N` just ok?

anangl replied:
I think `group` fits better.

gmarull commented:
Adding `DNM`, last commit should be dropped (increased buildkite capacity due to the large number of updated boards)

anangl commented on dts/bindings/pinctrl/nordic,nrf-pinctrl.yaml:
```diff
@@ -0,0 +1,107 @@
+          /* group 1 ('pins1' name is arbitrary) */
+          pins1 {
+            /* configure P0.1 as UART_TX and P0.2 as UART_RTS */
+            pins = <NRF_PSEL(UART_TX, 0, 1)>, <NRF_PSEL(UART_RTS, 0, 2)>;
```
```suggestion
            psels = <NRF_PSEL(UART_TX, 0, 1)>, <NRF_PSEL(UART_RTS, 0, 2)>;
```

anangl commented on dts/bindings/pinctrl/nordic,nrf-pinctrl.yaml:
```diff
@@ -0,0 +1,107 @@
+          /* group 2 */
+          pins2 {
+            /* configure P0.3 as UART_RX and P0.4 as UART_CTS */
+            pins = <NRF_PSEL(UART_RX, 0, 3)>, <NRF_PSEL(UART_CTS, 0, 4)>;
```
ditto

anangl commented on dts/bindings/pinctrl/nordic,nrf-pinctrl.yaml:
```diff
@@ -0,0 +1,107 @@
+    in a separate child node.
+
+    As shown, pin configurations are organized in groups within each child node.
+    Each group can specify a list of pin function selections in the 'pins' property.
```
```suggestion
    Each group can specify a list of pin function selections in the 'psels' property.
```

anangl commented on dts/bindings/pinctrl/nordic,nrf-pinctrl.yaml:
```diff
@@ -0,0 +1,107 @@
+
+       #include "board-pinctrl.dtsi"
+
+       uart0 {
```
```suggestion
       &uart0 {
```

anangl commented on dts/bindings/serial/nordic,nrf-uart-common.yaml:
```diff
@@ -21,59 +52,47 @@ properties:
         For pins P1.0 through P1.31, add 32 to the pin number. For
         example, to use P1.2 for TX, set:
 
-            tx-pin = <34>;  /* 32 + 2 */
+          tx-pin = <34>;  /* 32 + 2 */
```
Looks like an undesirable change.

anangl commented on include/dt-bindings/pinctrl/nrf-pinctrl.h:
```diff
@@ -0,0 +1,130 @@
+ *
+ * @param fun Pin function configuration (see NRF_FUNC_{name} macros).
+ * @param port Port (0 or 1).
+ * @param pin Pin (0..32).
```
```suggestion
 * @param pin Pin (0..31).
```

anangl commented on samples/boards/nrf/dynamic_pinctrl/sample.yaml:
```diff
@@ -0,0 +1,3 @@
+sample:
+  description: A dynamic pin control sample for nRF board/s.
+  name: dynamic_pinctrl
```
Shouldn't `nrf52840dk_nrf52840` be indicated as an integration platform?

gmarull replied:
added

anangl commented on samples/boards/nrf/dynamic_pinctrl/prj.conf:
```diff
@@ -0,0 +1,7 @@
+# Make sure GPIO is initialized before UART. This is required because we need
+# to read a GPIO pin to decide if UART pins have to be remapped or not, an
+# operation that needs to happen before UART device is initialized.
+CONFIG_GPIO_INIT_PRIORITY=50
+CONFIG_SERIAL_INIT_PRIORITY=70
+CONFIG_CONSOLE_INIT_PRIORITY=80
```
What about adding here also:
```
CONFIG_REMAP_INIT_PRIORITY=60
```

gmarull replied:
I've refactored this a little bit and added build assertion, please check again

anangl replied:
Yes, the build assertion is a good idea. And everything in this file is clear now. Thanks.

mbolivar-nordic commented on boards/arm/96b_carbon_nrf51/96b_carbon_nrf51-pinctrl.dtsi:
```diff
@@ -0,0 +1,22 @@
+&pinctrl {
+	uart0_default: uart0_default {
+		group1 {
+			psels = <NRF_PSEL(UART_TX, 0, 29)>,
+				<NRF_PSEL(UART_RX, 0, 11)>;
```
Did we ever settle on the above style versus this one?

```
			psels = <NRF_PSEL(UART_TX, 0, 29)
				 NRF_PSEL(UART_RX, 0, 11)>;
```

I seem to recall @carlescufi had an opinion on this, but I can't remember what it was.

gmarull replied:
I think nothing was agreed, but Linux seems to use

```
			psels = <NRF_PSEL(UART_TX, 0, 29)>,
				<NRF_PSEL(UART_RX, 0, 11)>;
```

mbolivar-nordic replied:
Just to be clear since I am going on vacation, I don't care personally. I just wanted to bring it up.

mbolivar-nordic commented on include/dt-bindings/pinctrl/nrf-pinctrl.h:
```diff
@@ -51,6 +51,15 @@
  * @{
  */
 
+/** UART TX */
```
That is, add a comment here saying "If you add functions here, make sure to keep the binding documentation for the psel property in sync"

gmarull replied:
I think we should just let Doxygen render dt-bindings public headers to avoid duplicate information/manual handling.

mbolivar-nordic replied:
How are users going to find that information from the bindings index, then?

gmarull replied:
`Refer to `include/dt-bindings/pinctrl/nrf-pinctrl.h for available functions`

mbolivar-nordic replied:
OK, I suppose.

gmarull commented on dts/bindings/serial/nordic,nrf-uart-common.yaml:
```diff
@@ -7,73 +7,92 @@ properties:
+    disable-rx:
+      type: boolean
+      description: |
+        Disable UART reception capabilities.
```
I have adjusted uart/e drivers to preserve old behavior, please check now. `CONFIG_PINCTRL=y` requires `disable-rx` though.

mbolivar-nordic replied:
> I have adjusted uart/e drivers to preserve old behavior

I will take your word for it. We can always open a bug if there's an issue.

gmarull commented on doc/guides/pinctrl/index.rst:
```diff
@@ -272,14 +272,14 @@ particular state are enclosed in a single Devicetree node.
         /* Node with pin configuration for default state */
         periph0_default: periph0_default {
-            pins1 {
+            group1 {
```
dropped commit

joerchan commented:
@gmarull Can this be used so that the application core can get the GPIO pins assigned to the network on nrf53?
The application core needs to know these pins because it has to grant the network core access to these pins. At the moment the UART pins are now hardcoded instead of from the device tree.
https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/nrf5340dk_nrf5340/nrf5340_cpunet_reset.c#L17

This becomes more complex when TFM is enabled since the MCU selection has to be done in the secure processing environment.

mbolivar-nordic reviewed (approved):
I would like to see https://github.com/zephyrproject-rtos/zephyr/pull/39429/commits/8e64dd0303e601b07608e44be0bcb0ec6cf5157c#r752354973 resolved at some point, but I don't consider it a blocker and it could be fixed in a follow-up PR.

I'm going on vacation so I'm going to approve this as from my perspective it looks great.

mbolivar-nordic commented:
> @gmarull Can this be used so that the application core can get the GPIO pins assigned to the network on nrf53?

This idea came from me, just for context. I was thinking maybe dynamic pin control could be used for this purpose from the board files. That would potentially let us put the pinctrl nodes into BOARD.dts and let the user override them.

joerchan commented:
The problem is really about getting the PIN assignment from the network core. For unique use-cases like MPSL FEM this is not so difficult, as the same DTS node can be included in both cores without issue. (for more context see here: https://github.com/zephyrproject-rtos/zephyr/pull/40513#issuecomment-975658625).
The problem with UART0 is that there is a node for it in both cores.
Just a quick look it looks like it might be easier to group the pin assignments here, but put them under a different node, like net-uart0 for example.

Sorry if I'm causing noise in this PR. We can start a discussion about this in another place.

carlescufi reviewed (approved):
Thank  you for all the effort put into this PR.
